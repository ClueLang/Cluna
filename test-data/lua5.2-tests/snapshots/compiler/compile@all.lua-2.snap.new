---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/all.lua
---
local version = "Lua 5.2"
if _VERSION ~= version {
    io.stderr::write("\nThis test suite is for ", version, ", not for ", _VERSION, "\nExiting tests\n");
    return 
}
_soft = false
_port = false
_no32 = false
_nomsg = false
_noposix = false
_nolonglong = false
_noformatA = false
local usertests = rawget(_G, "_U")
if usertests {
    _soft = true
    _port = true
    _no32 = true
    _nomsg = true
    _noposix = true
    _nolonglong = true
    _noformatA = true
}
if usertests {
    T = nil
}
T = rawget(_G, "T")
package.path = "?;./?.lua" .. package.path
math.randomseed(0);
collectgarbage("setstepmul", 200);
collectgarbage("setpause", 200);
print("current path:\n****" .. package.path .. "****\n");
local c = os.clock()
local collectgarbage = collectgarbage
{
    local msgs = {}
    global fn Message(m) {
        if !_nomsg {
            print(m);
            msgs[(#msgs + 1)] = string.sub(m, 3,  - 3)
        }
    }
    assert(os.setlocale("C"));
    local T, print, format, write, assert, type, unpack, floor = T, print, string.format, io.write, assert, type, table.unpack, math.floor
    local fn F(m) {
        local fn round(m) {
            m = m + 0.04999
            return m - (m % 0.1)
        }
        if m < 1000 {
            return m
        } else {
            m = m / 1000
            if m < 1000 {
                return round(m) .. "K"
            } else {
                return round(m / 1000) .. "M"
            }
        }
    }
    local showmem
    if !T {
        local max = 0
        showmem = fn () {
            local m = collectgarbage("count") * 1024
            max = (m > max) && m || max
            print(format("    ---- total memory: %s, max memory: %s ----\n", F(m), F(max)));
        }
    } else {
        showmem = fn () {
            T.checkmemory();
            local total, numblocks, maxmem = T.totalmem()
            local count = collectgarbage("count")
            print(format("\n    ---- total memory: %s (%.0fK), max use: %s,  blocks: %d\n", F(total), count, F(maxmem), numblocks));
            print(format("\t(strings:  %d, tables: %d, functions: %d, " .. "\n\tudata: %d, threads: %d)", T.totalmem("string"), T.totalmem("table"), T.totalmem("function"), T.totalmem("userdata"), T.totalmem("thread")));
        }
    }
    local fn report(n) {
        print("\n***** FILE '" .. n .. "'*****");
    }
    local olddofile = dofile
    dofile = fn (n) {
        showmem();
        report(n);
        local f = assert(loadfile(n))
        local b = string.dump(f)
        f = assert(load(b))
        return f()
    }
    dofile('main.lua');
    {
        local next, setmetatable, stderr = next, setmetatable, io.stderr
        local mt = {}
        mt.__gc = fn (o) {
            stderr::write('.');
            local n = setmetatable({}, mt)
            o = nil
            local a, b, c, d, e = nil
        }
        local n = setmetatable({}, mt)
    }
    report("gc.lua");
    local f = assert(loadfile('gc.lua'))
    f();
    collectgarbage("generational");
    dofile('db.lua');
    assert(dofile('calls.lua') == deep && deep);
    olddofile('strings.lua');
    olddofile('literals.lua');
    assert(dofile('attrib.lua') == 27);
    collectgarbage("incremental");
    olddofile('strings.lua');
    olddofile('literals.lua');
    dofile('constructs.lua');
    dofile('api.lua');
    collectgarbage("generational");
    collectgarbage("setpause", 200);
    collectgarbage("setmajorinc", 500);
    assert(dofile('locals.lua') == 5);
    dofile('constructs.lua');
    dofile('code.lua');
    if !_G._soft {
        report('big.lua');
        local f = coroutine.wrap(assert(loadfile('big.lua')))
        assert(f() == 'b');
        assert(f() == 'a');
    }
    dofile('nextvar.lua');
    dofile('pm.lua');
    dofile('api.lua');
    assert(dofile('events.lua') == 12);
    dofile('vararg.lua');
    dofile('closure.lua');
    dofile('coroutine.lua');
    dofile('goto.lua');
    dofile('errors.lua');
    dofile('math.lua');
    dofile('sort.lua');
    dofile('bitwise.lua');
    assert(dofile('verybig.lua') == 10);
    collectgarbage();
    dofile('files.lua');
    if #msgs > 0 {
        print("\ntests not performed:");
        for i = 1, #msgs {
            print(msgs[(i)]);
        }
        print();
    }
    print("final OK !!!");
    local debug = require("debug")
    debug.sethook(fn (a) {
        assert(type(a) == 'string');
    }, "cr");
    _G.showmem = showmem
}
local _G, showmem, print, format, clock, assert, open = _G, showmem, print, string.format, os.clock, assert, io.open
local fname = T && "time-debug.txt" || "time.txt"
local lasttime
if !usertests {
    local f = io.open(fname)
    if f {
        lasttime = assert(tonumber(f::read('*a')))
        f::close();
    } else {
        lasttime = nil
    }
}
print('cleaning all!!!!');
for n with pairs(_G) {
    if !({
        ___Glob = 1, 
        tostring = 1
    })[(n)] {
        _G[(n)] = nil
    }
}
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
showmem();
local time = clock() - c
print(format("\n\ntotal time: %.2f\n", time));
if !usertests {
    lasttime = lasttime || time
    local diff = (time - lasttime) / time
    local tolerance = 0.05
    assert(diff < tolerance && diff >  - tolerance);
    assert(open(fname, "w"))::write(time)::close();
}
