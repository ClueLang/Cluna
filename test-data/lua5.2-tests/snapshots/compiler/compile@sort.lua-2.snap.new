---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/sort.lua
---
print("testing (parts of) table library");
print("testing unpack");
local unpack = table.unpack
local x, y, z, a, n
a = {}
lim = 2000
for i = 1, lim {
    a[(i)] = i
}
assert(select(lim, unpack(a)) == lim && select('#', unpack(a)) == lim);
x = unpack(a)
assert(x == 1);
x = {
    unpack(a)
}
assert(#x == lim && x[(1)] == 1 && x[(lim)] == lim);
x = {
    unpack(a, lim - 2)
}
assert(#x == 3 && x[(1)] == lim - 2 && x[(3)] == lim);
x = {
    unpack(a, 10, 6)
}
assert(next(x) == nil);
x = {
    unpack(a, 11, 10)
}
assert(next(x) == nil);
x, y = unpack(a, 10, 10)
assert(x == 10 && y == nil);
x, y, z = unpack(a, 10, 11)
assert(x == 10 && y == 11 && z == nil);
a, x = unpack({
    1
})
assert(a == 1 && x == nil);
a, x = unpack({
    1, 
    2
}, 1, 1)
assert(a == 1 && x == nil);
if !_no32 {
    assert(!pcall(unpack, {}, 0, 2 ^ 31 - 1));
    assert(!pcall(unpack, {}, 1, 2 ^ 31 - 1));
    assert(!pcall(unpack, {},  - (2 ^ 31), 2 ^ 31 - 1));
    assert(!pcall(unpack, {},  - (2 ^ 31 - 1), 2 ^ 31 - 1));
    assert(pcall(unpack, {}, 2 ^ 31 - 1, 0));
    assert(pcall(unpack, {}, 2 ^ 31 - 1, 1));
    pcall(unpack, {}, 1, 2 ^ 31);
    a, b = unpack({
        2 ^ 31 - 1 = 20
    }, 2 ^ 31 - 1, 2 ^ 31 - 1)
    assert(a == 20 && b == nil);
    a, b = unpack({
        2 ^ 31 - 1 = 20
    }, 2 ^ 31 - 2, 2 ^ 31 - 1)
    assert(a == nil && b == 20);
}
print("testing pack");
a = table.pack()
assert(a[(1)] == nil && a.n == 0);
a = table.pack(table)
assert(a[(1)] == table && a.n == 1);
a = table.pack(nil, nil, nil, nil)
assert(a[(1)] == nil && a.n == 4);
print("testing sort");
local fn check(t) {
    local fn f(a, b) {
        assert(a && b);
        return true
    }
    local s, e = pcall(table.sort, t, f)
    assert(!s && e::find("invalid order function"));
}
check({
    1, 
    2, 
    3, 
    4
});
check({
    1, 
    2, 
    3, 
    4, 
    5
});
check({
    1, 
    2, 
    3, 
    4, 
    5, 
    6
});
global fn check(a, f) {
    f = f || fn (x, y) {
        return x < y
    }
    for n = #a, 2,  - 1 {
        assert(!f(a[(n)], a[(n - 1)]));
    }
}
a = {
    "Jan", 
    "Feb", 
    "Mar", 
    "Apr", 
    "May", 
    "Jun", 
    "Jul", 
    "Aug", 
    "Sep", 
    "Oct", 
    "Nov", 
    "Dec"
}
table.sort(a);
check(a);
global fn perm(s, n) {
    n = n || #s
    if n == 1 {
        local t = {
            unpack(s)
        }
        table.sort(t);
        check(t);
    } else {
        for i = 1, n {
            s[(i)], s[(n)] = s[(n)], s[(i)]
            perm(s, n - 1);
            s[(i)], s[(n)] = s[(n)], s[(i)]
        }
    }
}
perm({});
perm({
    1
});
perm({
    1, 
    2
});
perm({
    1, 
    2, 
    3
});
perm({
    1, 
    2, 
    3, 
    4
});
perm({
    2, 
    2, 
    3, 
    4
});
perm({
    1, 
    2, 
    3, 
    4, 
    5
});
perm({
    1, 
    2, 
    3, 
    3, 
    5
});
perm({
    1, 
    2, 
    3, 
    4, 
    5, 
    6
});
perm({
    2, 
    2, 
    3, 
    3, 
    5, 
    6
});
limit = 30000
if _soft {
    limit = 5000
}
a = {}
for i = 1, limit {
    a[(i)] = math.random()
}
local x = os.clock()
table.sort(a);
print(string.format("Sorting %d elements in %.2f sec.", limit, os.clock() - x));
check(a);
x = os.clock()
table.sort(a);
print(string.format("Re-sorting %d elements in %.2f sec.", limit, os.clock() - x));
check(a);
a = {}
for i = 1, limit {
    a[(i)] = math.random()
}
x = os.clock()
i = 0
table.sort(a, fn (x, y) {
    i = i + 1
    return y < x
});
print(string.format("Invert-sorting other %d elements in %.2f sec., with %i comparisons", limit, os.clock() - x, i));
check(a, fn (x, y) {
    return y < x
});
table.sort({});
for i = 1, limit {
    a[(i)] = false
}
x = os.clock()
table.sort(a, fn (x, y) {
    return nil
});
print(string.format("Sorting %d equal elements in %.2f sec.", limit, os.clock() - x));
check(a, fn (x, y) {
    return nil
});
for i, v with pairs(a) {
    assert(!v || i == 'n' && v == limit);
}
A = {
    "Ã¡lo", 
    "\0first :-)", 
    "alo", 
    "then this one", 
    "45", 
    "and a new"
}
table.sort(A);
check(A);
table.sort(A, fn (x, y) {
    load(string.format("A[%q] = ''", x))();
    collectgarbage();
    return x < y
});
tt = {
    __lt = fn (a, b) {
        return a.val < b.val
    }
}
a = {}
for i = 1, 10 {
    a[(i)] = {
        val = math.random(100)
    }
    setmetatable(a[(i)], tt);
}
table.sort(a);
check(a, tt.__lt);
check(a);
print("OK");
