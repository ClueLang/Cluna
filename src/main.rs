#![allow(non_snake_case)]
#![allow(non_upper_case_globals)]

macro_rules! check {
    ($tocheck: expr) => {
        match $tocheck {
            Ok(t) => t,
            Err(e) => return Err(e.to_string()),
        }
    };
}

macro_rules! arg {
    ($name: expr) => {
        unsafe { $name }
    };
}

// mod compiler;
mod compiler;
mod parser;
mod scanner;

use clap::Parser;
use compiler::compile_tokens;
use parser::*;
use scanner::*;
use std::{fs, fs::File, io::prelude::*, path::Path, time::Instant};

pub static mut ENV_TOKENS: bool = false;
pub static mut ENV_STRUCT: bool = false;
pub static mut ENV_OUTPUT: bool = false;
pub static mut ENV_DONTSAVE: bool = false;
pub static mut ENV_PATHISCODE: bool = false;
pub static mut ENV_NOMULTILINE: bool = false;

#[derive(Parser)]
#[clap(about, version, long_about = None)]
struct Cli {
    /// The path to the directory where the *.lua files are located.
    /// Every directory inside the given directory will be checked too.
    /// If the path points to a single *.lua file, only that file will be compiled.
    #[clap(required_unless_present = "license")]
    path: Option<String>,

    /// The name the output file will have
    #[clap(default_value = "main", value_name = "OUTPUT FILE NAME")]
    outputname: String,

    /// Print license information
    #[clap(short = 'L', long, display_order = 1000)]
    license: bool,

    /// Print list of detected tokens in compiled files
    #[clap(long)]
    tokens: bool,

    /// Print syntax structure of the tokens of the compiled files
    #[clap(long)]
    r#struct: bool,

    /// Print output Clue code in the console
    #[clap(short, long)]
    output: bool,

    /// Don't save compiled code
    #[clap(short = 'D', long)]
    dontsave: bool,

    /// Treat PATH not as a path but as clue code
    #[clap(short, long)]
    pathiscode: bool,

    /// Don't use multiline strings in the output
    #[clap(long)]
    nomultiline: bool,
}

fn compile_code(code: String, name: String, scope: usize) -> Result<String, String> {
    let time = Instant::now();
    let (tokens, _comments): (Vec<Token>, Vec<Comment>) = scan_code(code, name.clone())?;
    if arg!(ENV_TOKENS) {
        println!("Scanned tokens of file \"{}\":\n{:#?}", name, tokens);
    }
    let ctokens = parse_tokens(tokens, name.clone())?;
    if arg!(ENV_STRUCT) {
        println!("Parsed structure of file \"{}\":\n{:#?}", name, ctokens);
    }

    let code = compile_tokens(scope, ctokens);
    if arg!(ENV_OUTPUT) {
        println!("Compiled Lua code of file \"{}\":\n{}", name, code);
    }
    println!(
        "Compiled file \"{}\" in {} seconds!",
        name,
        time.elapsed().as_secs_f32()
    );
    Ok(code)
}

fn compile_file(path: &Path, name: String, scope: usize) -> Result<String, String> {
    let mut code: String = String::new();
    check!(check!(File::open(path)).read_to_string(&mut code));
    Ok(compile_code(code, name, scope)?)
}

fn compile_folder(path: &Path, rpath: String) -> Result<(), String> {
    for entry in check!(fs::read_dir(path)) {
        let entry = check!(entry);
        let name: String = entry
            .path()
            .file_name()
            .unwrap()
            .to_string_lossy()
            .into_owned();
        let filePathName: String = format!("{}/{}", path.display(), name);
        let filepath: &Path = &Path::new(&filePathName);
        let rname = rpath.clone() + &name;
        if filepath.is_dir() {
            compile_folder(filepath, rname + ".")?;
        } else if filePathName.ends_with(".lua") {
            let code = compile_file(filepath, name, 2)?;
            let compiled_path = format!(
                "{}{}",
                filepath.display().to_string().strip_suffix(".lua").unwrap(),
                ".clue"
            );
            if !arg!(ENV_DONTSAVE) {
                check!(fs::write(compiled_path, code))
            }
        }
    }
    Ok(())
}

fn main() -> Result<(), String> {
    let cli = Cli::parse();
    if cli.license {
        println!("{}", include_str!("../LICENSE"));
        return Ok(());
    }
    unsafe {
        ENV_TOKENS = cli.tokens;
        ENV_STRUCT = cli.r#struct;
        ENV_OUTPUT = cli.output;
        ENV_DONTSAVE = cli.dontsave;
        ENV_PATHISCODE = cli.pathiscode;
        ENV_NOMULTILINE = cli.nomultiline;
    }
    let codepath = cli.path.unwrap();
    if arg!(ENV_PATHISCODE) {
        println!(
            "{}",
            compile_code(codepath.clone(), "(command line)".to_owned(), 0)?
        );
        return Ok(());
    }
    let path: &Path = Path::new(&codepath);

    if path.is_dir() {
        compile_folder(path, String::new())?;
    } else if path.is_file() {
        let compiled_name =
            String::from(path.display().to_string().strip_suffix(".lua").unwrap()) + ".clue";
        let code = compile_file(
            path,
            path.file_name().unwrap().to_string_lossy().into_owned(),
            0,
        )?;

        if !arg!(ENV_DONTSAVE) {
            check!(fs::write(compiled_name, code))
        }
    } else {
        return Err(String::from("The given path doesn't exist"));
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use crate::compile_folder;
    use std::path::Path;

    #[test]
    fn compile() {
        compile_folder(Path::new("tests/"), String::new()).unwrap();
    }
}
