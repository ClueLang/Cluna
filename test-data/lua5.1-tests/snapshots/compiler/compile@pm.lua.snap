---
source: src/main.rs
expression: compiled
input_file: test-data/lua5.1-tests/pm.lua
---
print('testing pattern matching');
global fn f(s, p) {
    local i, e = string.find(s, p)
    if i {
        return string.sub(s, i, e)
    }
}
global fn f1(s, p) {
    p = string.gsub(p, "%%([0-9])", fn (s) {
        return "%" .. (s + 1)
    })
    p = string.gsub(p, "^(^?)", "%1()", 1)
    p = string.gsub(p, "($?)$", "()%1", 1)
    local t = {
        string.match(s, p)
    }
    return string.sub(s, t[(1)], t[(#t)] - 1)
}
a, b = string.find('', '')
assert(a == 1 && b == 0);
a, b = string.find('alo', '')
assert(a == 1 && b == 0);
a, b = string.find('a\0o a\0o a\0o', 'a', 1)
assert(a == 1 && b == 1);
a, b = string.find('a\0o a\0o a\0o', 'a\0o', 2)
assert(a == 5 && b == 7);
a, b = string.find('a\0o a\0o a\0o', 'a\0o', 9)
assert(a == 9 && b == 11);
a, b = string.find('a\0a\0a\0a\0\0ab', '\0ab', 2)
assert(a == 9 && b == 11);
a, b = string.find('a\0a\0a\0a\0\0ab', 'b')
assert(a == 11 && b == 11);
assert(string.find('a\0a\0a\0a\0\0ab', 'b\0') == nil);
assert(string.find('', '\0') == nil);
assert(string.find('alo123alo', '12') == 4);
assert(string.find('alo123alo', '^12') == nil);
assert(f('aloALO', '%l*') == 'alo');
assert(f('aLo_ALO', '%a*') == 'aLo');
assert(f('aaab', 'a*') == 'aaa');
assert(f('aaa', '^.*$') == 'aaa');
assert(f('aaa', 'b*') == '');
assert(f('aaa', 'ab*a') == 'aa');
assert(f('aba', 'ab*a') == 'aba');
assert(f('aaab', 'a+') == 'aaa');
assert(f('aaa', '^.+$') == 'aaa');
assert(f('aaa', 'b+') == nil);
assert(f('aaa', 'ab+a') == nil);
assert(f('aba', 'ab+a') == 'aba');
assert(f('a$a', '.$') == 'a');
assert(f('a$a', '.%$') == 'a$');
assert(f('a$a', '.$.') == 'a$a');
assert(f('a$a', '$$') == nil);
assert(f('a$b', 'a$') == nil);
assert(f('a$a', '$') == '');
assert(f('', 'b*') == '');
assert(f('aaa', 'bb*') == nil);
assert(f('aaab', 'a-') == '');
assert(f('aaa', '^.-$') == 'aaa');
assert(f('aabaaabaaabaaaba', 'b.*b') == 'baaabaaabaaab');
assert(f('aabaaabaaabaaaba', 'b.-b') == 'baaab');
assert(f('alo xo', '.o$') == 'xo');
assert(f(' \n isto � assim', '%S%S*') == 'isto');
assert(f(' \n isto � assim', '%S*$') == 'assim');
assert(f(' \n isto � assim', '[a-z]*$') == 'assim');
assert(f('um caracter ? extra', '[^%sa-z]') == '?');
assert(f('', 'a?') == '');
assert(f('�', '�?') == '�');
assert(f('�bl', '�?b?l?') == '�bl');
assert(f('  �bl', '�?b?l?') == '');
assert(f('aa', '^aa?a?a') == 'aa');
assert(f(']]]�b', '[^]]') == '�');
assert(f("0alo alo", "%x*") == "0a");
assert(f("alo alo", "%C+") == "alo alo");
print('+');
assert(f1('alo alx 123 b\0o b\0o', '(..*) %1') == "b\0o b\0o");
assert(f1('axz123= 4= 4 34', '(.+)=(.*)=%2 %1') == '3= 4= 4 3');
assert(f1('=======', '^(=*)=%1$') == '=======');
assert(string.match('==========', '^([=]*)=%1$') == nil);
local fn range(i, j) {
    if i <= j {
        return i, range(i + 1, j)
    }
}
local abc = string.char(range(0, 255))
assert(string.len(abc) == 256);
global fn strset(p) {
    local res = {
        s = ''
    }
    string.gsub(abc, p, fn (c) {
        res.s = res.s .. c
    });
    return res.s
}
assert(string.len(strset('[\200-\210]')) == 11);
assert(strset('[a-z]') == "abcdefghijklmnopqrstuvwxyz");
assert(strset('[a-z%d]') == strset('[%da-uu-z]'));
assert(strset('[a-]') == "-a");
assert(strset('[^%W]') == strset('[%w]'));
assert(strset('[]%%]') == '%]');
assert(strset('[a%-z]') == '-az');
assert(strset('[%^%[%-a%]%-b]') == '-[]^ab');
assert(strset('%Z') == strset('[\1-\255]'));
assert(strset('.') == strset('[\1-\255%z]'));
print('+');
assert(string.match("alo xyzK", "(%w+)K") == "xyz");
assert(string.match("254 K", "(%d*)K") == "");
assert(string.match("alo ", "(%w*)$") == "");
assert(string.match("alo ", "(%w+)$") == nil);
assert(string.find("(�lo)", "%(�") == 1);
local a, b, c, d, e = string.match("�lo alo", "^(((.).).* (%w*))$")
assert(a == '�lo alo' && b == '�l' && c == '�' && d == 'alo' && e == nil);
a, b, c, d = string.match('0123456789', '(.+(.?)())')
assert(a == '0123456789' && b == '' && c == 11 && d == nil);
print('+');
assert(string.gsub('�lo �lo', '�', 'x') == 'xlo xlo');
assert(string.gsub('alo �lo  ', ' +$', '') == 'alo �lo');
assert(string.gsub('  alo alo  ', '^%s*(.-)%s*$', '%1') == 'alo alo');
assert(string.gsub('alo  alo  \n 123\n ', '%s+', ' ') == 'alo alo 123 ');
t = "ab� d"
a, b = string.gsub(t, '(.)', '%1@')
assert('@' .. a == string.gsub(t, '', '@') && b == 5);
a, b = string.gsub('ab�d', '(.)', '%0@', 2)
assert(a == 'a@b@�d' && b == 2);
assert(string.gsub('alo alo', '()[al]', '%1') == '12o 56o');
assert(string.gsub("abc=xyz", "(%w*)(%p)(%w+)", "%3%2%1-%0") == "xyz=abc-abc=xyz");
assert(string.gsub("abc", "%w", "%1%0") == "aabbcc");
assert(string.gsub("abc", "%w+", "%0%1") == "abcabc");
assert(string.gsub('���', '$', '\0��') == '���\0��');
assert(string.gsub('', '^', 'r') == 'r');
assert(string.gsub('', '$', 'r') == 'r');
print('+');
assert(string.gsub("um (dois) tres (quatro)", "(%(%w+%))", string.upper) == "um (DOIS) tres (QUATRO)");
{
    local fn setglobal(n, v) {
        rawset(_G, n, v);
    }
    string.gsub("a=roberto,roberto=a", "(%w+)=(%w%w*)", setglobal);
    assert(_G.a == "roberto" && _G.roberto == "a");
}
global fn f(a, b) {
    return string.gsub(a, '.', b)
}
assert(string.gsub("trocar tudo em |teste|b| � |beleza|al|", "|([^|]*)|([^|]*)|", f) == "trocar tudo em bbbbb � alalalalalal");
local fn dostring(s) {
    return loadstring(s)() || ""
}
assert(string.gsub("alo $a=1$ novamente $return a$", "$([^$]*)%$", dostring) == "alo  novamente 1");
x = string.gsub("$x=string.gsub('alo', '.', string.upper)$ assim vai para $return x$", "$([^$]*)%$", dostring)
assert(x == ' assim vai para ALO');
t = {}
s = 'a alo jose  joao'
r = string.gsub(s, '()(%w+)()', fn (a, w, b) {
    assert(string.len(w) == b - a);
    t[(a)] = b - a
})
assert(s == r && t[(1)] == 1 && t[(3)] == 3 && t[(7)] == 4 && t[(13)] == 4);
global fn isbalanced(s) {
    return string.find(string.gsub(s, "%b()", ""), "[()]") == nil
}
assert(isbalanced("(9 ((8))(\0) 7) \0\0 a b ()(c)() a"));
assert(notisbalanced("(9 ((8) 7) a b (\0 c) a"));
assert(string.gsub("alo 'oi' alo", "%b''", '"') == 'alo " alo');
local t = {
    "apple", 
    "orange", 
    "lime", 
    n = 0
}
assert(string.gsub("x and x and x", "x", fn () {
    t.n = t.n + 1
    return t[(t.n)]
}) == "apple and orange and lime");
t = {
    n = 0
}
string.gsub("first second word", "%w%w*", fn (w) {
    t.n = t.n + 1
    t[(t.n)] = w
});
assert(t[(1)] == "first" && t[(2)] == "second" && t[(3)] == "word" && t.n == 3);
t = {
    n = 0
}
assert(string.gsub("first second word", "%w+", fn (w) {
    t.n = t.n + 1
    t[(t.n)] = w
}, 2) == "first second word");
assert(t[(1)] == "first" && t[(2)] == "second" && t[(3)] == nil);
assert(notpcall(string.gsub, "alo", "(.", print));
assert(notpcall(string.gsub, "alo", ".)", print));
assert(notpcall(string.gsub, "alo", "(.", {}));
assert(notpcall(string.gsub, "alo", "(.)", "%2"));
assert(notpcall(string.gsub, "alo", "(%1)", "a"));
assert(notpcall(string.gsub, "alo", "(%0)", "a"));
local a = string.rep('a', 300000)
assert(string.find(a, '^a*.?$'));
assert(notstring.find(a, '^a*.?b$'));
assert(string.find(a, '^a-.?$'));
global fn rev(s) {
    return string.gsub(s, "(.)(.+)", fn (c, s1) {
        return rev(s1) .. c
    })
}
local x = string.rep('012345', 10)
assert(rev(rev(x)) == x);
assert(string.gsub("alo alo", ".", {}) == "alo alo");
assert(string.gsub("alo alo", "(.)", {
    a = "AA", 
    l = ""
}) == "AAo AAo");
assert(string.gsub("alo alo", "(.).", {
    a = "AA", 
    l = "K"
}) == "AAo AAo");
assert(string.gsub("alo alo", "((.)(.?))", {
    al = "AA", 
    o = false
}) == "AAo AAo");
assert(string.gsub("alo alo", "().", {
    2, 
    5, 
    6
}) == "256 alo");
t = {}
setmetatable(t, {
    __index = fn (t, s) {
        return string.upper(s)
    }
});
assert(string.gsub("a alo b hi", "%w%w+", t) == "a ALO b HI");
assert(string.gfind == string.gmatch);
local a = 0
for i with string.gmatch('abcde', '()') {
    assert(i == a + 1);
    a = i
}
assert(a == 6);
t = {
    n = 0
}
for w with string.gmatch("first second word", "%w+") {
    t.n = t.n + 1
    t[(t.n)] = w
}
assert(t[(1)] == "first" && t[(2)] == "second" && t[(3)] == "word");
t = {
    3, 
    6, 
    9
}
for i with string.gmatch("xuxx uu ppar r", "()(.)%2") {
    assert(i == table.remove(t, 1));
}
assert(table.getn(t) == 0);
t = {}
for i, j with string.gmatch("13 14 10 = 11, 15= 16, 22=23", "(%d+)%s*=%s*(%d+)") {
    t[(i)] = j
}
a = 0
for k, v with pairs(t) {
    assert(k + 1 == v + 0);
    a = a + 1
}
assert(a == 3);
assert(string.gsub("aaa aa a aaa a", "%f[%w]a", "x") == "xaa xa x xaa x");
assert(string.gsub("[[]] [][] [[[[", "%f[[].", "x") == "x[]] x]x] x[[[");
assert(string.gsub("01abc45de3", "%f[%d]", ".") == ".01abc.45de.3");
assert(string.gsub("01abc45 de3x", "%f[%D]%w", ".") == "01.bc45 de3.");
assert(string.gsub("function", "%f[\1-\255]%w", ".") == ".unction");
assert(string.gsub("function", "%f[^\1-\255]", ".") == "function.");
local i, e = string.find(" alo aalo allo", "%f[%S].-%f[%s].-%f[%S]")
assert(i == 2 && e == 5);
local k = string.match(" alo aalo allo", "%f[%S](.-%f[%s].-%f[%S])")
assert(k == 'alo ');
local a = {
    1, 
    5, 
    9, 
    14, 
    17
}
for k with string.gmatch("alo alo th02 is 1hat", "()%f[%w%d]") {
    assert(table.remove(a, 1) == k);
}
assert(table.getn(a) == 0);
print('OK');
